[MRG+2] In roc_curve: drop some suboptimal thresholds

gclenaghan opened the pull request:
Currently, the roc_curve function returns a value for every distinct possible threshold. For many classification routines this is close to the number of data points, and the results of this can get very unwieldy. For instance, with a large test set, the plotted ROC curve is unsuitable for a vector graphics format since the image file will be large and slow to load. Moreover, many of the thresholds are suboptimal: one could choose a different threshold for a higher TPR at the same FPR or a lower FPR at the same TPR.

I've added an option keep_all_thresholds to the roc_curve function which drops thresholds which are invisible on a plotted ROC curve when set to false (by default it is set to true to ensure compatibility). Basically, in the sorted list of scores, if two of the same class appear consecutively, it will drop threshold corresponding to the first.

Here are some tests:

```
In [21]:
t1, t2
import sklearn.datasets
import sklearn.metrics as met
import time
probas_pred, y_true = sklearn.datasets.make_classification(n_samples=1000000,
                                    n_features=1, n_informative=1,
                                    n_redundant=0, n_clusters_per_class=1,
                                    random_state=1031)
​
t0 = time.time()
fpr, tpr, thr = met.roc_curve(y_true, probas_pred)
t1 = time.time() - t0
​
t0 = time.time()
fpr2, tpr2, thr2 = met.roc_curve(y_true, probas_pred, keep_all_thresholds=False)
t2 = time.time() - t0
​
t1, t2

Out[21]:
(0.2661151885986328, 0.24276185035705566)

In [18]:

fpr.shape, fpr2.shape

Out[18]:
((990748,), (9843,))

In [19]:

met.auc(fpr, tpr) - met.auc(fpr2, tpr2)

Out[19]:
-2.2459999415858078e-09
```

With this option roc_curve is faster and returns many fewer results (how many fewer depends on how good the classifier is, results may vary).

I've also changed some tests to use the new option.

I've been doing something similar to this with the results of roc_curve but it would be quicker and easier to have it directly in the function. Please let me know if there is anything I forgot to do!

gclenaghan committed:
Implement dropping of suboptimal thresholds in roc_curve

jnothman commented:
I'm fine with this, even as the default setting, in principle.

I'd like to see the logic moved out of `_binary_clf_curve` to `roc_curve`. It's not applicable to `precision_recall_curve`. If controlled by a parameter, perhaps rename it to `drop_redundant` or `drop_intermediate`.

gclenaghan commented:
Thanks for the feedback!

I placed the computation in `_binary_clf_curve` as it avoids any unnecessary repeats of computations, in particular it's important that the sorting of `y_true` matches that of `tps` and friends, so to do this in `roc_curve` would mean sorting a second time.

I'm less familiar with how precision-recall curves are used so I didn't add this, but the same thresholds which are optimal on a ROC curve should be optimal for precision/recall, the main difference is that the interpolation between suboptimal points is no longer linear. So for plotting you might want more points, but for analysis (computing the highest precision for a given recall, etc.) the filtered set returned by this addition should be enough.

Does this seem like a reasonable choice? Let me know what you think. I can certainly change the name of the parameter.

jnothman commented:
Have you actually found the calculation of the ROC curve to be a bottleneck? I suspect a 10% speed-up is not helpful, but a summarised curve might be useful for users and reduce plot-related costs (e.g. file size for vector graphics). So don't worry about a little bit of repeated work; you should be able to derive the selected indices from tps without needing to sort again.

gclenaghan commented:
Good point, I didn't think to use the tps to find the indices. I will write something up.

gclenaghan commented:
Okay, now the logic is in roc_curve. It's slightly slower (on the same 1 million point data set above, it takes 0.28s) but you were right and this is very reasonable.

Let me know if there are any other tweaks necessary!

jnothman commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -126,14 +126,16 @@ def _average_precision(y_true, y_score):
-    assert_equal(fpr.shape, tpr.shape)
-    assert_equal(fpr.shape, thresholds.shape)
+
+    for drop in [True, False]:
```
This test tells us nothing if that particular case didn't need to drop anything. Ensure that something is actually dropped, and then hopefully asserting the equivalence of AUC should suffice as a test.

jnothman commented:
I'd be happy for this to become the default, or even make the prior behaviour unavailable, but I'd like to hear from others. @arjoly?

arjoly commented:
Can you explain more why we shouldn't expect a modification of the auc in all cases? Otherwise said, why are those points sub-optimal? Is there a weird extreme case where it should be a problem?

I am bit wary as it could have side effects such the added isclose ( which is now a problem).

jnothman commented:
Basically, where the number of true positives changes at threshold t_i
identically to at t_{i-1} then only one of tpr or fpr is affected: if it's
another true example, tpr increases and fpr stays constant; another false
example means fpr increases while tpr is constant. These are points
strictly within horizontals and verticals on the ROC curve, so their
vertices are not visible, nor do they affect the AUC.

On 24 September 2015 at 18:28, Arnaud Joly <notifications@github.com> wrote:

> Can you explain more why we shouldn't expect a modification of the auc in
> all cases? Otherwise said, why are those points sub-optimal? Is there a
> weird extreme case where it should be a problem?
>
> I am bit wary as it could have side effects such the added isclose ( which
> is now a problem).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/5237#issuecomment-142856343>
> .
>

jnothman commented:
What I wrote wasn't quite correct, rather:

At any change of threshold, only one of tpr or fpr is affected: if it's another true example, tpr increases and fpr stays constant; another false example means fpr increases while tpr is constant. That is, the roc curve is made up of horizontal and vertical segments. We only need to keep vertices, being where the curve changes from a horizontal to a vertical segment or vv.

gclenaghan commented:
@arjoly: This only drops points on the curve which are in between and collinear with points which are left in. As such, the curve itself, if drawn with linear interpolation between the points, is identical, so the area is also identical (up to floating point arithmetic... this takes many less operations to compute though). A point is suboptimal if one can achieve higher tpr at the same fpr or lower fpr at the same tpr: there is no reason to choose the threshold associated with that point. The only points which are dropped satisfy one of these conditions, but the step pattern is preserved so they appear identical (actually you would not want to choose the threshold associated with the bottom of a step, but this is kept in to preserve the shape of the curve). I do not believe there are any problematic cases, and I've seen other implementations do something similar (I think this was an internal script someone wrote so it's not something I can point you to, unfortunately).

@jnothman This isn't true when there are multiple y_true's with the same y_score, in which case roc_curve (or rather _binary_clf_curve) drops them already, creating a point where both fps and tps may increase from the previous, a diagonal line on the curve. This is why in my code I check if both of them increase, and still drop points which are collinear with their surrounding points. I'm writing some additional tests now as you suggest and I'll be sure to test this case.

gclenaghan commented:
Okay, I've added a new test to ensure it is dropping the thresholds I intended it to, including a case where there are several repeated thresholds. In cases like this any rational slope is possible and you can think of what it's doing with np.diff(_,2) as taking the second derivative and checking if the curvature is zero: if so it's collinear and dropped. In the included test case, just looking at one of tps or fps would erroneously drop a point.

arjoly commented:
Thanks for the explication and also for the tests! Can you transform this knowledge as code comment.

gclenaghan commented:
Sure, I modified the comment to be more verbose.

gclenaghan commented:
Are there any other changes that would help? Let me know what I can do!

jnothman commented:
LGTM

arjoly commented:
LGTM

jnothman commented:
@gclenaghan do you mind squashing your commits into one to avoid the numerous merge artefacts? at the same time, please add a note in doc/whats_new.rst about the enhancement. thanks!

gclenaghan commented:
Okay, done. Thanks!

jnothman commented:
Thank you!
